File: libcomponent/src/main/java/com/bear/libcomponent/component/ComponentContainer.java
package com.bear.libcomponent.component;

import java.util.Map;

public class ComponentContainer {

    private final RootComponent rootComponent = new RootComponent();

    public <C extends GroupComponent> void regComponent(C component, Object tag) {
        if (component == null) {
            return;
        }
        rootComponent.regComponent(component, tag);
    }

    public <C extends GroupComponent> C getComponent(Class<C> clz, Object tag) {
        return rootComponent.getComponent(clz, tag);
    }

    public Map<ComponentKey<?>, GroupComponent> getComponentMap() {
        return rootComponent.getComponentMap();
    }

    public <C extends IComponent> boolean contain(Class<C> clz, Object tag) {
        Map<ComponentKey<?>, GroupComponent> componentMap = getComponentMap();
        if (!componentMap.isEmpty()) {
            ComponentKey<?> componentKey = new ComponentKey<>(clz, tag);
            return componentMap.containsKey(componentKey);
        }
        return false;
    }

    private static class RootComponent extends GroupComponent {

    }
}

File: libcomponent/src/main/java/com/bear/libcomponent/core/ComponentContainer.kt
package com.bear.libcomponent.core

import com.bear.libcomponent.component.base.ComponentKey
import com.bear.libcomponent.component.base.GroupComponent

internal class ComponentContainer {
    val rootComponent = RootComponent()

    val componentMap: Map<ComponentKey<*>, IComponent>
        get() = rootComponent.componentMap

    fun <C : IComponent> regComponent(component: C, tag: Any? = null) {
        rootComponent.regComponent(component, tag)
    }

    fun <C : IComponent> getComponent(clz: Class<C>, tag: Any? = null): C? {
        return rootComponent.getComponent(clz, tag)
    }

    fun <C : IComponent> contain(clz: Class<C>, tag: Any? = null): Boolean {
        return rootComponent.getComponent(clz, tag) != null
    }
}

internal class RootComponent : GroupComponent()